pallets/name-service/src/lib.rs

```diff
@@ -0,0 +1,414 @@
+        ) -> DispatchResult {
+            let _who = ensure_signed(origin)?;
+
+            Self::do_accept_register(&username, registrant)?;
+
+            Ok(())
+        }
+
+        /// Set a registered and owned username as Primary
+        #[pallet::call_index(2)]
+        #[pallet::weight(T::WeightInfo::set_primary_name())]
+        #[transactional]
+        pub fn set_primary_name(
+            origin: OriginFor<T>,
+            username: UserName,
+            registrant: ZkAddressType,
+        ) -> DispatchResult {
+            let _who = ensure_signed(origin)?;
+
+            Self::try_set_primary_name(username, registrant)?;
+
+            Ok(())
+        }
+
+        /// Cancel pending name for register
+        #[pallet::call_index(3)]
+        #[pallet::weight(T::WeightInfo::cancel_pending_register())]
+        #[transactional]
+        pub fn cancel_pending_register(
+            origin: OriginFor<T>,
+            username: UserName,
+            registrant: ZkAddressType,
+        ) -> DispatchResult {
+            let _who = ensure_signed(origin)?;
+
+            Self::try_cancel_pending_register(username, registrant)?;
+
+            Ok(())
+        }
+
+        /// Remove Already Registered Name
+        #[pallet::call_index(4)]
+        #[pallet::weight(T::WeightInfo::remove_register())]
+        #[transactional]
+        pub fn remove_register(
+            origin: OriginFor<T>,
+            username: UserName,
+            registrant: ZkAddressType,
+        ) -> DispatchResult {
+            let _who = ensure_signed(origin)?;
+
+            Self::try_remove_register(username, registrant)?;
+
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Returns the account ID of this pallet.
+    #[inline]
+    pub fn account_id() -> T::AccountId {
+        T::PalletId::get().into_account_truncating()
+    }
+
+    /// Queue username for regiser
+    fn do_register(username: &UserName, registrant: ZkAddressType) -> DispatchResult {
+        // Username checks
+        username_validation(username).ok_or(Error::<T>::InvalidUsernameFormat)?;
+
+        let (hash_user, hash_address) = (
+            T::Hashing::hash_of(username),
+            T::Hashing::hash_of(&registrant),
+        );
+
+        // Check if already Pending Register
+        ensure!(
+            !PendingRegister::<T>::contains_key(hash_user),
+            Error::<T>::AlreadyPendingRegister
+        );
+
+        // Check if already registered
+        ensure!(
+            !UsernameRecords::<T>::contains_key(username),
+            Error::<T>::NameAlreadyRegistered
+        );
+
+        PendingRegister::<T>::insert(
+            hash_user,
+            (
+                hash_address,
+                frame_system::Pallet::<T>::block_number()
+                    .saturating_add(T::RegisterWaitingPeriod::get()),
+            ),
+        );
+
+        Self::deposit_event(Event::NameQueuedForRegister);
```

Comment:
Can we add some field into Event?

Reply:
Can you expand? Pending Names are Hashed so they cant be queried, so that information would be useless